File: Project_Sourcecode_java/src/main/java/edu/iit/cs586/project/input/Account2.java
package edu.iit.cs586.project.input;

import edu.iit.cs586.project.MDA_EFSM;
import edu.iit.cs586.project.abstractfactory.AbstractFactory;
import edu.iit.cs586.project.datastore.DS2;
import edu.iit.cs586.project.datastore.DataStore;

/**
 * @author CharmingLee <charminglee@jieztech.com>
 * @date 2022/5/1 20:28
 */
public class Account2 {
    //m is a pointer to the MDA-EFSM object
    private MDA_EFSM m;
    //ds is a pointer to the Data Store DS-2 object In the data store:
    private DataStore ds;

    public Account2 (AbstractFactory af, MDA_EFSM m) {
        this.ds = af.getDataStore();
        this.m = m;
    }

    public void OPEN(int p, int y, float a) {
        ((DS2) ds).temp_p = p;
        ((DS2) ds).temp_y = y;
        ((DS2) ds).temp_a = a;

        m.Open();
    }

    public void PIN(int x) {
        if (x == ds.pn) {
            m.CorrectPinAboveMin();
        } else {
            m.IncorrectPin(2);
        }
    }

    public void DEPOSIT(float d) {
        ((DS2) ds).temp_d = d;
        m.Deposit();
    }

    public void WITHDRAW(float w) {
        ((DS2) ds).temp_w = w;
        if (((DS2) ds).b > 0) {
            m.Withdraw();
            m.AboveMinBalance();
        } else {
            m.NoFunds();
        }
    }

    public void BALANCE() {
        m.Balance();
    }

    public void LOGIN(int y) {
        if (y == ds.id) {
            m.Login();
        } else {
            m.IncorrectLogin();
        }
    }

    public void LOGOUT() {
        m.Logout();
    }

    public void suspend() {
        m.Suspend();
    }

    public void activate() {
        m.Activate();
    }

    public void close() {
        m.Close();
    }
}

File: Project_Sourcecode_go/client/Account2.go
package client

import (
	"cs586_project/abstractfactory"
	"cs586_project/datastore"
	"cs586_project/state"
)

type Account2 struct {
	m  *state.MdaEfsm
	ds interface{}
}

func (a2 *Account2) Init(m *state.MdaEfsm, factory abstractfactory.AbstractFactoryApi) {
	a2.m = m
	a2.ds = factory.GetDs()
}

func (a2 *Account2) OPEN(p int, y int, a float64) {
	ds1 := a2.ds.(*datastore.DS2)
	ds1.Temp_p = p
	ds1.Temp_y = y
	ds1.Temp_a = a

	a2.m.Open()
}

func (a2 *Account2) PIN(x int) {
	ds := a2.ds.(*datastore.DS2)
	if x == ds.Pin {
		if ds.Balance > 100 {
			a2.m.CorrectPinAboveMin()
		} else {
			a2.m.CorrectPinBelowMin()
		}
	} else {
		a2.m.IncorrectPin(1)
	}
}

func (a2 *Account2) DEPOSIT(d float64) {
	ds := a2.ds.(*datastore.DS2)
	ds.Temp_d = d
	a2.m.Deposit()
}

func (a2 *Account2) WITHDRAW(w float64) {
	ds := a2.ds.(*datastore.DS2)
	ds.Temp_w = w
	if ds.Balance > 0 {
		a2.m.Withdraw()
		a2.m.AboveMinBalance()
	} else {
		a2.m.NoFunds()
	}
}

func (a2 *Account2) BALANCE() {
	a2.m.Balance()
}

func (a2 *Account2) LOGIN(y int) {
	ds := a2.ds.(*datastore.DS2)
	if y == ds.Id {
		a2.m.Login()
	} else {
		a2.m.IncorrectLogin()
	}
}

func (a2 *Account2) LOGOUT() {
	a2.m.Logout()
}

func (a2 *Account2) Suspend() {
	a2.m.Suspend()
}

func (a2 *Account2) Activate() {
	a2.m.Activate()
}

func (a2 *Account2) Close() {
	a2.m.Close()
}
